src/wayland/dmabuf/mod.rs

```diff
@@ -0,0 +1,465 @@
+            params.failed();
+        }
+    }
+
+    fn create_immed(
+        &mut self,
+        params: BufferParams,
+        buffer_id: NewResource<wl_buffer::WlBuffer>,
+        width: i32,
+        height: i32,
+        format: u32,
+        flags: u32,
+    ) {
+        // Cannot reuse a params:
+        if self.used {
+            params.as_ref().post_error(
+                ParamError::AlreadyUsed as u32,
+                "This buffer_params has already been used to create a buffer.".into(),
+            );
+            return;
+        }
+        self.used = true;
+        if !buffer_basic_checks(
+            &self.formats,
+            &self.pending_planes,
+            &params,
+            format,
+            width,
+            height,
+        ) {
+            trace!(self.log, "Killing client providing bogus dmabuf buffer params.");
+            return;
+        }
+        let info = BufferInfo {
+            planes: ::std::mem::replace(&mut self.pending_planes, Vec::new()),
+            width,
+            height,
+            format,
+            flags,
+        };
+        let mut handler = self.handler.borrow_mut();
+        if let Ok(data) = handler.validate_dmabuf(info) {
+            trace!(self.log, "Creating a new validated immediate dma wl_buffer.");
+            handler.create_buffer(data, buffer_id);
+        } else {
+            trace!(
+                self.log,
+                "Refusing creation of an invalid immediate dma wl_buffer, killing client."
+            );
+            params.as_ref().post_error(
+                ParamError::InvalidWlBuffer as u32,
+                "create_immed resulted in an invalid buffer.".into(),
+            );
+        }
+    }
+}
+
+fn buffer_basic_checks(
+    formats: &[Format],
+    pending_planes: &[Plane],
+    params: &BufferParams,
+    format: u32,
+    width: i32,
+    height: i32,
+) -> bool {
+    // protocol_checks:
+    // This must be a known format
+    let format = match formats.iter().find(|f| f.format == format) {
+        Some(f) => f,
+        None => {
+            params.as_ref().post_error(
+                ParamError::InvalidFormat as u32,
+                format!("Format {:x} is not supported.", format),
+            );
+            return false;;
+        }
+    };
+    // The number of planes set must match what the format expects
+    let max_plane_set = pending_planes.iter().map(|d| d.plane_idx + 1).max().unwrap_or(0);
+    if max_plane_set != format.plane_count || pending_planes.len() < format.plane_count as usize {
+        params.as_ref().post_error(
+            ParamError::Incomplete as u32,
+            format!(
+                "Format {:x} requires {} planes but got {}.",
+                format.format, format.plane_count, max_plane_set
+            ),
+        );
+        return false;
+    }
+    // Width and height must be positivie
+    if width < 1 || height < 1 {
```

Comment:
Does it need to be strictly greater? Isn't a 1x1 buffer valid as well?

Reply:
This code allows a 1x1 buffer, it errors if either width or height is strictly smaller than 1